video/out/drm_atomic.c

```diff
@@ -458,3 +485,93 @@ bool drm_mode_destroy_blob(int fd, struct drm_mode *mode)
+
+    struct hdr_output_metadata metadata = {0};
+    metadata.metadata_type = DRM_STATIC_METADATA_TYPE1;
+    metadata.hdmi_metadata_type1.metadata_type = DRM_STATIC_METADATA_TYPE1;
+
+    hdr_data->current_csp = hdr_data->new_csp;
+    struct mp_colorspace *csp = &hdr_data->current_csp;
+    struct mp_csp_primaries colors = mp_get_csp_primaries(csp->primaries);
+
+    switch (csp->gamma) {
+    case MP_CSP_TRC_BT_1886:
+        metadata.hdmi_metadata_type1.eotf = DRM_METADATA_EOTF_TRADITIONAL_HDR;
+        break;
+    case MP_CSP_TRC_PQ:
+        metadata.hdmi_metadata_type1.eotf = DRM_METADATA_EOTF_SMPTE_2084;
+        break;
+    case MP_CSP_TRC_HLG:
+        metadata.hdmi_metadata_type1.eotf = DRM_METADATA_EOTF_HLG;
+        break;
+    default:
+        // Default is SDR
+        metadata.hdmi_metadata_type1.eotf = DRM_METADATA_EOTF_TRADITIONAL_SDR;
+    }
+
+    if (metadata.hdmi_metadata_type1.eotf != DRM_METADATA_EOTF_TRADITIONAL_SDR) {
+        // Convert to nits (cd/m2):
+        float min_luminance = 0.0;
+        float max_luminance = csp->sig_peak * MP_REF_WHITE;
+        float max_cll = max_luminance;
+        float max_fall = max_luminance;
+ 
+        // 0x0000 to 0xC350 maps to 0.0 - 1.0000. Step of 0.00002
+        metadata.hdmi_metadata_type1.display_primaries[0].x = clamp16_round(colors.red.x);
+        metadata.hdmi_metadata_type1.display_primaries[0].y = clamp16_round(colors.red.y);
+        metadata.hdmi_metadata_type1.display_primaries[1].x = clamp16_round(colors.green.x);
+        metadata.hdmi_metadata_type1.display_primaries[1].y = clamp16_round(colors.green.y);
+        metadata.hdmi_metadata_type1.display_primaries[2].x = clamp16_round(colors.blue.x);
+        metadata.hdmi_metadata_type1.display_primaries[2].y = clamp16_round(colors.blue.y);
+
+        // White point, same mapping as display primaries
+        metadata.hdmi_metadata_type1.white_point.x = clamp16_round(colors.white.x);
+        metadata.hdmi_metadata_type1.white_point.y = clamp16_round(colors.white.y);
+
+        // Min Luminance: 16-bit value in units of 0.0001 cd/m2
+        // 0x0001 represents 0.0001 cd/m2 and 0xFFFF represents 6.5535 cd/m2.
+        metadata.hdmi_metadata_type1.min_display_mastering_luminance = clamp16(floorf(min_luminance * 10000.0));
+
+        // Max Luminance: 16-bit value in units of 1 cd/m2
+        // 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2.
+        metadata.hdmi_metadata_type1.max_display_mastering_luminance = clamp16(ceilf(max_luminance));
+
+        // Max Content Light Level: 16-bit value in units of 1 cd/m2
+        // 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2.
+        metadata.hdmi_metadata_type1.max_cll = clamp16(ceilf(max_cll));
+
+        // Max Frame Average Light Level: 16-bit value in units of 1 cd/m2
+        // 0x0001 represents 1 cd/m2 and 0xFFFF represents 65535 cd/m2.
+        metadata.hdmi_metadata_type1.max_fall = clamp16(ceilf(max_fall));
```

Comment:
I'm not exactly an expert on the Linux DRM subsystem, but it seems strange that this is being calculated for any non-SDR mode while it is only applicable to the various PQ-based standards (HDR10/HDR10+/DV) and not to HLG.

Per CTA 861.G - "If a Source does not have SMPTE ST 2086 [41] metadata to send, then Data Bytes 3-22 shall be
populated with zeroes to indicate the data is not provided. " which appears to me to be the most sensible route for HLG since HLG content fundamentally does not have ST.2086 metadata.

Reply:
Indeed most of what is done in this function is not correct, the actual "HDR" data being used isn't even correct AFAICT. I do plan on finishing this PR, or at the very least rebasing it so it stays current, but there are a few other things that need to be done before this PR can be finished

1. A general DRM refactor, given the discussion above as well as my own thoughts on the DRM code.
2. A general infrastructure for parsing HDR data out of files and passing it to functions like this. That will also help when wayland gets HDR and theoretically it may help macOS support, too.